File: Linear/Dynamic/Linked List/Circular Linked List/02_Deleting/Given_Position/Deleting.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def deleteAtEnd(last, pos):
    head = last.next

    if (last is None or pos < 1) or (head is last or pos is 1):
        return None
    
    if pos is 1:
        last.next = head.next
        return last
    
    prev = head
    curr = head.next

    for i in range(2, pos, 1):
        if curr is head:
            return last
        prev = curr
        curr = curr.next
    
    if curr is head:
        return last
    
    prev.next = curr.next

    if curr is last:
        last = prev

    return last

def printList(last):
    if last is None:
        return
    
    head = last.next
    temp = head

    while True:
        print(temp.data, end="")
        temp = temp.next
        if temp is not head:
            print(" -> ", end="")
        else:
            break
    
    print()

# Create circular linked list: 2 -> 3 -> 4
first = Node(2)
first.next = Node(3)
first.next.next = Node(4)

last = first.next.next
last.next = first

print("Linked List before deletion\n")
printList(last)

pos = int(input("Enter a position: \n"))
last = deleteAtEnd(last, pos)

print("Linked List after deletion\n")
printList(last)

File: Linear/Dynamic/Linked List/Circular Linked List/02_Deleting/Given_Position/Deleting.js
import readline from 'readline'

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

function deleteAtGivenPosition(last, pos) {
    let head = last.next;

    if ((last === null || pos < 1) || (head === last || pos === 1))
        return null;

    if (pos === 1) {
        last.next = head.next;
        return last;
    }

    let prev = head;
    let curr = head.next;

    for (let i = 2; i < pos && curr !== head; i++) {
        prev = curr;
        curr = curr.next;
    }

    if (curr === head)
        return last;

    prev.next = curr.next;

    if (curr === last)
        last = prev;

    return last;
}

function printList(last) {
    if (last === null)
        return;

    let head = last.next;
    let temp = head;

    do {
        process.stdout.write(temp.data.toString())
        temp = temp.next;
        if (temp !== head)
            process.stdout.write(" -> ");
    } while (temp !== head);

    console.log();
}

// Create circular linked list: 2 -> 3 -> 4
let first = new Node(2);
first.next = new Node(3);
first.next.next = new Node(4);

let last = first.next.next;
last.next = first;

console.log("Linked List before deletion");
printList(last);

rl.question("Enter a position: ", (ans) => {
    const pos = parseInt(ans)

    last = deleteAtGivenPosition(last, pos);
    
    console.log("Linked List after deletion");
    printList(last);

    rl.close()
})